### PR TITLE
chore: add key/value schema name prop and refactor

### PR DESCRIPTION
**Pull request by lihaosky**
### Description 
* Add `KEY_SCHEMA_FULL_NAME` and `VALUE_SCHEMA_FULL_NAME` to set schema name explicitly for avro, protobuf and json_sr format
* Add properties for protobuf, json, json_sr and introduce base class `ConnectProperties`
* Refactor `key_schema_id`, `value_schema_id` in `ConnectFormat` to be just `schema_id` in `ConnectProperties` since it exists separately in key and value format properties so using same name is fine.

### Testing done 
Unit test

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/properties/with/CreateSourceAsPropertiesTest.java [hunk 1]
```diff
@@ -149,10 +154,37 @@ public void shouldNotSetAvroSchemaNameForNonAvroKey() {
...
+            .put(CommonCreateConfigs.VALUE_FORMAT_PROPERTY, new StringLiteral("Protobuf"))
+            .put(CommonCreateConfigs.VALUE_SCHEMA_FULL_NAME, new StringLiteral("schema"))
+            .build());
+
+    // Then:
+    assertThat(properties.getValueFormatProperties().get(ConnectProperties.FULL_SCHEMA_NAME), is("schema"));
+  }
+
+  @Test
+  public void shouldSetFullKeySchemaName() {
```
Comment (mjsax) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/properties/with/CreateSourceAsPropertiesTest.java:177]: nit: `shouldSetKeyFullSchemaName()`

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/properties/with/CreateSourceAsPropertiesTest.java [hunk 2]
```diff
@@ -149,10 +154,37 @@ public void shouldNotSetAvroSchemaNameForNonAvroKey() {
...
+    // Given:
+    final CreateSourceAsProperties props = CreateSourceAsProperties
+        .from(ImmutableMap.<String, Literal>builder()
+            .put(FORMAT_PROPERTY, new StringLiteral("Json_sr"))
+            .put(KEY_SCHEMA_FULL_NAME, new StringLiteral("KeySchema"))
+            .build());
+
+    // When / Then:
```
Comment (mjsax) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/properties/with/CreateSourceAsPropertiesTest.java:185]: nit: Why `When / Then`  and not just `Then` ?

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/properties/with/CreateSourceAsPropertiesTest.java [hunk 3]
```diff
@@ -149,10 +154,37 @@ public void shouldNotSetAvroSchemaNameForNonAvroKey() {
...
+    assertThat(props.getKeyFormatProperties("json_sr", "foo"),
+        hasEntry(ConnectProperties.FULL_SCHEMA_NAME, "KeySchema"));
```
Comment (mjsax) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/properties/with/CreateSourceAsPropertiesTest.java:187]: nit: why `hasEntry` compared to `...get(...), is(...)` as above?

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonFormat.java [hunk 4]
```diff
@@ -55,6 +55,11 @@ protected ConnectSchemaTranslator getConnectSchemaTranslator(
     throw new UnsupportedOperationException(name() + " does not implement Schema Registry support");
   }
 
+  @Override
+  public Set<String> getSupportedProperties() {
+    return JsonProperties.SUPPORTED_PROPERTIES;
```
Comment (mjsax) [inline ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonFormat.java:60]: Why this change? I though JSON won't support this, only JSON_SR ?

Might be easier to just `return ImmutableSet.of();` ?

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonProperties.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+package io.confluent.ksql.serde.json;
+
+import com.google.common.collect.ImmutableSet;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.confluent.ksql.serde.connect.ConnectProperties;
+import java.util.Map;
+
+public class JsonProperties extends ConnectProperties {
+
+  static final ImmutableSet<String> SUPPORTED_PROPERTIES = ImmutableSet.of();
```
Comment (mjsax) [inline ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonProperties.java:25]: Why do we need to introduce this class if the set of supported properties is empty?

File: ksqldb-serde/src/test/java/io/confluent/ksql/serde/json/JsonPropertiesTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    // When:
+    final Exception e = assertThrows(UnsupportedOperationException.class,
+        properties::getFullSchemaName);
+
+    // Then:
+    assertThat(e.getMessage(), is("JSON does not implement Schema Registry support"));
+  }
+
+  @Test
+  public void shouldGetDefaultFullSchemaName() {
```
Comment (mjsax) [inline ksqldb-serde/src/test/java/io/confluent/ksql/serde/json/JsonPropertiesTest.java:58]: Seem this test is a duplicate of the previous one?

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonSchemaProperties.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.serde.json;
+
+import com.google.common.collect.ImmutableSet;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.util.Map;
+
+public class JsonSchemaProperties extends JsonProperties {
```
Comment (mjsax) [inline ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonSchemaProperties.java:22]: Should we directly `extend ConnectProperties` ? Not sure if having `JsonProperties` at all make sense?

File: ksqldb-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateSourceFactory.java [hunk 8]
```diff
@@ -229,7 +229,12 @@ private Formats buildFormats(
       final KsqlConfig ksqlConfig
   ) {
     final FormatInfo keyFormat = SourcePropertiesUtil.getKeyFormat(props, name);
+    // Validate format name and property
+    FormatFactory.of(keyFormat);
```
[the remark this answers is not in the document]
Comment (lihaosky) [reply ksqldb-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateSourceFactory.java:233]: Not sure exactly. Might be validated previously in other places like `DefaultSchemaInjector` and `SchemaRegisterInjector` implicitly. I don't fully recall why I added this, might be because I wrote some unit test for other PR previously and found incompatible properties can be set... 

Should be good to have a central place to do all validations explicitly so the preconditions are clear. 

I'm ok to remove this given this could be checked in other code path.

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/properties/with/CreateSourceAsPropertiesTest.java (hunk 3 above)
Comment (lihaosky) [reply ksqldb-parser/src/test/java/io/confluent/ksql/parser/properties/with/CreateSourceAsPropertiesTest.java:187]: Good catch. Don't recall this. Will change previous one to `hasEntry`

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonFormat.java (hunk 4 above)
Comment (lihaosky) [reply ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonFormat.java:60]: Because I introduced `JsonProperties`, so make it consistent with other format properties class to have `SUPPORTED_PROPERTIES` field. So we can modify `SUPPORTED_PROPERTIES` when new properties are added later.

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonProperties.java (hunk 5 above)
Comment (lihaosky) [reply ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonProperties.java:25]: 1. Json extended `ConnectFormat`, so I thought it might make sense to create `JsonProperties` extends `ConnectProperties`?
2. Both Json and JsonSchema format call https://github.com/confluentinc/ksql/blob/master/ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonSerdeFactory.java#L40 to create serializer atm. I would still like to maintain the behavior and pass a property object of Json or JsonSchema to that class later. That's why I create `JsonProperties` and let `JsonSchemaProperties` extend it so that I can pass both `JsonProperties` and `JsonSchemaProperties` to `KsqlJsonSerdeFactory`

That being said, I guess 2 is a stronger reason but there should be other ways to solve it as well.

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonSchemaProperties.java (hunk 7 above)
Comment (lihaosky) [reply ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/JsonSchemaProperties.java:22]: See comment above, I think extends `ConnectProperties` is fine.

**Review by mjsax — APPROVED** [review]
